src/main/java/apoc/algo/pagerank/PageRankArrayStorageParallelCypher.java

```diff
@@ -0,0 +1,324 @@
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.kernel.impl.core.ThreadToStatementContextBridge;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.Arrays;
+import java.util.concurrent.Future;
+
+import static apoc.algo.pagerank.PageRankUtils.toFloat;
+import static apoc.algo.pagerank.PageRankUtils.toInt;
+
+public class PageRankArrayStorageParallelCypher implements PageRank
+{
+    public static final int ONE_MINUS_ALPHA_INT = toInt( ONE_MINUS_ALPHA );
+    public static final int WRITE_BATCH=10_100;
+    public final int BATCH_SIZE = 100_000;
+    private final GraphDatabaseAPI db;
+    private final ExecutorService pool;
+    private String nodeCypher;
+    private String relCypher;
+    private int nodeCount;
+    private int relCount;
+
+
+    /*
+        1. Memory usage right now:
+            6 arrays : size of nodes.
+            2 arrays : size of relationships.
+            Memory usage: 6*N + 2*M
+        2. TODO:
+        Run this in parallel using ExecutorService.
+        Sort and divide into (number of threads) chunks and process it in parallel.
+     */
+
+    int [] nodeMapping;
+    int [] sourceDegreeData;
+
+    // Weighted Degrees.
+    // Don't absolutely need following two. Can be calculated on the fly.
+    int [] sourceWeightData;
+    int [] sourceChunkStartingIndex;
+
+    // Storing relationships
+    int [] relationshipTarget;
+    int [] relationshipWeight;
+
+    // Output arrays.
+    int [] pageRanks;
+    int [] previousPageRanks;
+
+    public PageRankArrayStorageParallelCypher(
+            GraphDatabaseAPI db,
+            ExecutorService pool,
+            String nodeCypher,
+            String relCypher)
+    {
+        this.pool = pool;
+        this.db = db;
+        this.relCypher = relCypher;
+        this.nodeCypher = nodeCypher;
+        readDataIntoArray(this.relCypher, this.nodeCypher);
+    }
+
+    private int getNodeIndex(int node) {
+        int index = Arrays.binarySearch(nodeMapping, node);
+        return index;
+    }
+
+    private int getDegree(int node) {
+        int degree = 0;
+        int index = Arrays.binarySearch(nodeMapping, node);
+        if (index >= 0) {
+            degree = sourceDegreeData[index];
+        }
+        return degree;
+    }
+
+    private void setDegree(int node, int degree) {
+        int index = Arrays.binarySearch(nodeMapping, node);
+        sourceDegreeData[index] = degree;
+    }
+
+    private int getWeight(int node) {
+        int degree = 0;
+        int index = Arrays.binarySearch(nodeMapping, node);
+        if (index >= 0) {
+            degree = sourceWeightData[index];
+        }
+        return degree;
+    }
+
+    private void setWeight(int node, int weight) {
+        int index = Arrays.binarySearch(nodeMapping, node);
+        sourceWeightData[index] = weight;
+    }
+
+    private void calculateChunkIndices() {
```

Comment:
This was one doubt that I had earlier. As far as I think, unless the user has actually sorted the rel_cypher according to the source id, we might have relationships that are random:
source target weight
3           4           1
1           4           1
3          1            3

So in this case, when I get the first relationship at 3, I should be knowing where in the array I should be storing the targetid and weight. So for that I'm iterating over the relationship cypher twice. In the first iteration, I calculate the degrees of all the nodes. Then this function calculated the projected starting chunkIndices where the chunk for a given node would be stored and then in the next iteration over the cypher results, I actually store the target and store.